simapp/mint_fn.go

```diff
@@ -0,0 +1,131 @@
+		}
+
+		var stakingPool stakingtypes.QueryPoolResponse
+		err = env.QueryRouterService.InvokeTyped(ctx, &stakingtypes.QueryPoolRequest{}, &stakingPool)
+		if err != nil {
+			return err
+		}
+
+		// bondedRatio
+		bondedRatio := math.LegacyNewDecFromInt(stakingPool.Pool.BondedTokens).QuoInt(stakingTokenSupply.Amount)
+		minter.Inflation = minter.NextInflationRate(mintParams.Params, bondedRatio)
+		minter.AnnualProvisions = minter.NextAnnualProvisions(mintParams.Params, stakingTokenSupply.Amount)
+
+		// to get a more accurate amount of tokens minted, we get, and later store, last minting time.
+
+		// if this is the first time minting, we initialize the minter.Data with the current time - 60s
+		// to mint tokens at the beginning. Note: this is a custom behavior to avoid breaking tests.
+		if minter.Data == nil {
+			minter.Data = make([]byte, 8)
+			binary.BigEndian.PutUint64(minter.Data, (uint64)(env.HeaderService.HeaderInfo(ctx).Time.Unix()-60))
+		}
+
+		lastMint := binary.BigEndian.Uint64(minter.Data)
+		binary.BigEndian.PutUint64(minter.Data, (uint64)(env.HeaderService.HeaderInfo(ctx).Time.Unix()))
+
+		// calculate the amount of tokens to mint, based on the time since the last mint
+		secsSinceLastMint := env.HeaderService.HeaderInfo(ctx).Time.Unix() - (int64)(lastMint)
+		provisionAmt := minter.AnnualProvisions.QuoInt64(31536000).MulInt64(secsSinceLastMint) // 31536000 = seconds in a year
+		mintedCoin := sdk.NewCoin(mintParams.Params.MintDenom, provisionAmt.TruncateInt())
+		maxSupply := mintParams.Params.MaxSupply
+		totalSupply := stakingTokenSupply.Amount
+
+		if !maxSupply.IsZero() {
+			// supply is not infinite, check the amount to mint
+			remainingSupply := maxSupply.Sub(totalSupply)
+
+			if remainingSupply.LTE(math.ZeroInt()) {
+				// max supply reached, no new tokens will be minted
+				// also handles the case where totalSupply > maxSupply
+				return nil
+			}
+
+			// if the amount to mint is greater than the remaining supply, mint the remaining supply
+			if mintedCoin.Amount.GT(remainingSupply) {
+				mintedCoin.Amount = remainingSupply
+			}
+		}
+
+		if mintedCoin.Amount.IsZero() {
+			// skip as no coins need to be minted
+			return nil
+		}
+
+		mintedCoins := sdk.NewCoins(mintedCoin)
+		if err := bankKeeper.MintCoins(ctx, minttypes.ModuleName, mintedCoins); err != nil {
+			return err
+		}
+
+		// Example of custom send while minting
+		// Send some tokens to a "team account"
+		// if err = bankKeeper.SendCoinsFromModuleToAccount(ctx, minttypes.ModuleName, ... ); err != nil {
+		// 	return err
+		// }
+
+		// TODO: figure how to get FeeCollectorName from mint module without generating a cyclic dependency
+		if err = bankKeeper.SendCoinsFromModuleToModule(ctx, minttypes.ModuleName, authtypes.FeeCollectorName, mintedCoins); err != nil {
+			return err
+		}
+
+		if mintedCoin.Amount.IsInt64() {
+			defer telemetry.ModuleSetGauge(minttypes.ModuleName, float32(mintedCoin.Amount.Int64()), "minted_tokens")
+		}
```

Comment:
id avoid the extra telemetry. This can be calculated via queries.

Reply:
removed